packages/circuits/tests/check-substring-match.test.ts

```diff
@@ -0,0 +1,104 @@
+import { wasm as wasm_tester } from "circom_tester";
+import path from "path";
+
+describe("CheckSubstringMatch Circuit", () => {
+    let circuit: any;
+
+    beforeAll(async () => {
+        circuit = await wasm_tester(
+            path.join(
+                __dirname,
+                "./test-circuits/check-substring-match-test.circom"
+            ),
+            {
+                recompile: true,
+                include: path.join(__dirname, "../../../node_modules"),
+                output: path.join(__dirname, "./compiled-test-circuits"),
+            }
+        );
+    });
+
+    it("should match when substring is at the beginning", async () => {
+        const input = {
+            in: [1, 2, 3, 4, 5, ...Array(27).fill(0)],
+            substring: [1, 2, 3, ...Array(29).fill(0)],
+        };
+        const witness = await circuit.calculateWitness(input);
+        await circuit.checkConstraints(witness);
+        expect(witness[1]).toBe(1n);
+    });
+
+    it("should not match when substring is different", async () => {
+        const input = {
+            in: [1, 2, 3, 4, 5, ...Array(27).fill(0)],
+            substring: [1, 2, 4, ...Array(29).fill(0)],
+        };
+        const witness = await circuit.calculateWitness(input);
+        await circuit.checkConstraints(witness);
+        expect(witness[1]).toBe(0n);
+    });
+
+    it("should match with full length substring", async () => {
+        const input = {
+            in: Array(32)
+                .fill(0)
+                .map((_, i) => i + 1),
+            substring: Array(32)
+                .fill(0)
+                .map((_, i) => i + 1),
+        };
+        const witness = await circuit.calculateWitness(input);
+        await circuit.checkConstraints(witness);
+        expect(witness[1]).toBe(1n);
+    });
+
+    it("should fail when first element of substring is zero", async () => {
+        const input = {
+            in: [1, 2, 3, 4, 5, ...Array(27).fill(0)],
+            substring: [0, 2, 3, ...Array(29).fill(0)],
+        };
+        await expect(async () => {
+            await circuit.calculateWitness(input);
+        }).rejects.toThrow("Assert Failed");
+    });
+
+    it("should not match when substring is not at the beginning", async () => {
```

Comment:
Wait what, it should match right?

Reply:
This is test if for the helper circuit CheckSubstringMatch which is used to check for matches from each starting index - so I think it correct.
Btw, why cant we reveal data by taking a startIndex and length as input? I think that would be still cheaper. Is there a use-case we need to enforce there is only one substring?

Reply:
Thank you for answering that @saleel, and you are correct. `CheckSubstringMatch` checks only a part of the input (basically a window), so this template is actually used in `CountSubstringOccurrences`.

> Btw, why cant we reveal data by taking a startIndex and length as input? I think that would be still cheaper. Is there a use-case we need to enforce there is only one substring?

Yes, there is a use-case where the json data might be like:
```json
{
    command: "do this",
    other_string: "command: \"do this instead\""
}
```

And we want to reveal `n` characters after the substring `command`. Someone can spoof it like this.